### PR TITLE
feat: add custom reducers to estimators [DET-3098]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

Introduce custom reducers for estimator trial.  From the docstring in the PR:

    During distributed evaluation, many types of metrics calculated via ``tf.metrics`` or
    ``tf.keras.metrics`` cannot be aggregated properly from the per-slot final metrics
    calculated by each separate Estimator replica. One example is ``tf.metrics.auc``, where
    the ROC AUC calculated over predictions and labels from a full dataset cannot be derived
    from a set of ROC AUC metrics evaluated over the shards of a dataset. However, with
    ``make_metric``, the ROC AUC could be calculated in distributed training by calling
    ``sklearn.metrics.roc_auc_score`` in a custom ``reducer`` function.

## Test Plan

Lots of manual testing, in addition to adding a new unit test and a new parallel test.

### CODE REVIEW

File: harness/determined/estimator/_estimator_context.py [hunk 1]
```diff
@@ -142,3 +144,26 @@ def allgather_metrics(self, metrics: Any) -> List:
         if self._allgather_fn is None:
             raise AssertionError("allgather_metrics must not be called before training begins")
         return self._allgather_fn(metrics)
+
+    def _build_allgather_op(self, f: Callable) -> Callable:
+        """Build an op that uses allgather in a way that is safely sequentialized."""
+
+        def decorated(*arg: Any, **kwarg: Any) -> tf.Operation:
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_context.py:151]: non-blocking: seems like we are wrapping only internal functions so shouldn't matter that much, but doesn't hurt to add in `functools.wraps()` here.

File: harness/determined/estimator/_estimator_context.py [hunk 2]
```diff
@@ -142,3 +144,26 @@ def allgather_metrics(self, metrics: Any) -> List:
...
+                new_op = f(*arg, **kwarg)
+            self._allgather_ops.append(new_op)
+            return new_op
+
+        return decorated
+
+    def make_metric(
+        self,
+        metric: Any,
+        reducer: Union[Callable[[List[Any]], Any], "estimator.MetricReducer"],
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_context.py:162]: question: why is `estimator.MetricReducer` in quotes?

File: harness/determined/estimator/_reducer.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        (sum, count) tuples and it would calculate the overall mean.
+
+            def cross_slot_reduce(self, per_slot_metrics):
+                sums, counts = zip(*per_slot_metrics)
+                return np.array(sum(sums) / sum(counts))
+        """
+        pass
+
+
+class _SimpleMetricReducer(MetricReducer):
```
Comment (aaron276h) [inline harness/determined/estimator/_reducer.py:79]: This might change after you add in documentation, but my initial impression is that having this SimpleReducer that takes a arbitrary `reduce` function presents so many different options for the API that it becomes confusing. All of this is generally targeted at an advanced use case, and asking them to define their own `MetricReducer` seems ok.

File: harness/determined/estimator/_reducer.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        across slots, so accumulate will be called many times on each slot.
+
+        accumulate should return the accumulated state.  After evaluation is complete, the final
+        return value of accumulate will become an element of the per_slot_metrics argument to
+        cross_slot_reduce.
+
+        In the example of the calculating a distributed mean, accumulate might keep a running sum
+        and a count of metrics received:
+
+            def accumulate(self, metric):
```
Comment (aaron276h) [inline harness/determined/estimator/_reducer.py:54]: praise: really nice informative doc strings

File: harness/determined/estimator/_estimator_context.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_estimator_context.py:162]: It is a forward reference, like it is not yet defined when this portion of the code is read by the interpreter.

File: harness/determined/estimator/_reducer.py (hunk 4 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_reducer.py:54]: thank you!

File: harness/determined/estimator/_reducer.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_reducer.py:79]: I think they both add a lot of value.  I expounded on the difference in one of the new docstrings I just added.

File: harness/determined/estimator/_estimator_context.py (hunk 1 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_estimator_context.py:151]: good idea

File: harness/determined/estimator/_estimator_context.py [hunk 5]
```diff
@@ -133,3 +133,98 @@ class EstimatorExperimentalContext(_data_layer.DataLayerContext):
...
+        name: str = "custom_metric",
+    ) -> tf.keras.metrics.Metric:
+        """
+        Return an estimator-compatible metric which will be calculated properly, even during
+        distributed training.
+
+        Unfortunately, during distributed evaluation, many types of metrics calculated via
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_context.py:171]: Drop `Unfortunately`

File: harness/determined/estimator/_estimator_context.py [hunk 6]
```diff
@@ -133,3 +133,98 @@ class EstimatorExperimentalContext(_data_layer.DataLayerContext):
...
+
+                   return tf.estimator.EstimatorSpec(
+                       mode,
+                       loss=loss,
+                       eval_metric_ops={"my_avg_prediction": my_avg_prediction},
+                   )
+
+        The primary motivation for passing a function as the reducer is simplicity.  Metrics from
+        all batches will be buffered in memory and passed over the network where they will be
+        reduced all at once.  This is probably fine for scalar metrics, or on validation datasets
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_context.py:212]: Rephrase: `This is probably fine`

File: harness/determined/estimator/_estimator_context.py [hunk 7]
```diff
@@ -142,3 +145,86 @@ def allgather_metrics(self, metrics: Any) -> List:
...
+                        metric=predictions, reducer=my_mean_reducer
+                   )
+
+                   return tf.estimator.EstimatorSpec(
+                       mode,
+                       loss=loss,
+                       eval_metric_ops={"my_avg_prediction": my_avg_prediction},
+                   )
+
+        The primary motivation for passing a function as the reducer is simplicity.  Metrics from
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_context.py:210]: it would be more clear imo if this came right after the paragraph that describes the two options.

File: harness/determined/estimator/_estimator_context.py (hunk 6 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_estimator_context.py:216]: rewrote the whole second half of this paragraph, let me know if you like it.

File: harness/determined/estimator/_estimator_context.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_estimator_context.py:175]: done

File: harness/determined/estimator/_estimator_context.py (hunk 7 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_estimator_context.py:214]: done.

**Comment by rb-determined-ai** [comment]
@aaron276h This is now ready for a "for-realsies" review.  The incremental update since the last time you looked is:

 - added unit and parallel e2e test
 - fixed a bug where the `tf.control_dependencies()` was somehow triggering the call to `_DistributedMetric.result()` twice, resulting in building two allgather ops for every metric in the graph.  I fixed this by pre-building every result op and returning the cached op when `result()` is called.
 - the `_allgather_ops` list in the context has to be reset after every `.evaluate()` call, because every `.evaluate()` is going to build a clean graph, and you don't want to call `tf.control_dependencies()` on ops from the old graph.

File: e2e_tests/tests/fixtures/estimator_dataset/model.py [hunk 8]
```diff
@@ -31,14 +31,35 @@
...
+        self.sum += metric.sum()
+        return self.sum
 
-class EstimatorDebugTrial(EstimatorTrial):
-    def __init__(self, context: EstimatorTrialContext):
+    def cross_slot_reduce(self, per_slot_metrics: List):
+        return sum(per_slot_metrics)
+
+
+class EstimatorDebugTrial(estimator.EstimatorTrial):
```
Comment (aaron276h) [inline e2e_tests/tests/fixtures/estimator_dataset/model.py:61]: non-blocking: can we rename this from Debug to something else

File: harness/determined/estimator/_estimator_context.py [hunk 9]
```diff
@@ -142,3 +145,92 @@ def allgather_metrics(self, metrics: Any) -> List:
...
+        strategy may also be desirable for quick prototyping or for calculating metrics that are
+        difficult or impossible to calculate incrementally.
+
+        The primary motivation for passing a ``det.estimator.MetricsReducer`` as the reducer is
+        performance. ``det.estimator.MetricsReducer`` allows the user to incrementally calculate
+        the partial metric on each slot, taking advantage of distributed computation, minimizing
+        memory usage, and minimizing the network communication before the final
+        ``cross_slot_reduce`` operation.
+
+        In either the single-function reducer case or the ``det.estimator.MetricsReducer`` case,
```
Comment (aaron276h) [inline harness/determined/estimator/_estimator_context.py:205]: I couldn't really understand this part

File: harness/determined/estimator/_reducer.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
+import abc
+from typing import Any, Callable, List
+
+import numpy as np
+import tensorflow as tf
+
+from determined import estimator
+
+
+class MetricReducer:
```
Comment (aaron276h) [inline harness/determined/estimator/_reducer.py:10]: Question: do we want to show this in the docs as well?

File: harness/tests/experiment/fixtures/estimator_linear_model.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+    def accumulate(self, metric: Any):
+        self.sum += metric.sum()
+        return self.sum
+
+    def cross_slot_reduce(self, per_slot_metrics: List):
+        return sum(per_slot_metrics)
+
+
+class LinearEstimator(estimator.EstimatorTrial):
```
Comment (aaron276h) [inline harness/tests/experiment/fixtures/estimator_linear_model.py:43]: praise: nice tests

File: e2e_tests/tests/experiment/test_tf_estimator.py [hunk 12]
```diff
@@ -150,6 +150,28 @@ def test_mnist_estimator_data_layer_lfs(tf2: bool) -> None:
     run_mnist_estimator_data_layer_test(tf2, "lfs")
 
 
+@pytest.mark.parallel  # type: ignore
+@pytest.mark.parametrize("tf2", [True, False])  # type: ignore
+def test_custom_reducer_distributed(secrets: Dict[str, str], tf2: bool) -> None:
+    config = conf.load_config(conf.fixtures_path("estimator_dataset/distributed.yaml"))
+    # Run with multiple steps to verify we are resetting reducers right.
+    config = conf.set_max_steps(config, 2)
+    config = conf.set_slots_per_trial(config, 2)
```
Comment (aaron276h) [inline e2e_tests/tests/experiment/test_tf_estimator.py:159]: blocking: this experiment will never get scheduled since this needs to be 8 slots.

File: e2e_tests/tests/experiment/test_tf_estimator.py [hunk 13]
```diff
@@ -150,6 +150,28 @@ def test_mnist_estimator_data_layer_lfs(tf2: bool) -> None:
...
+    config = conf.set_tf2_image(config) if tf2 else conf.set_tf1_image(config)
+
+    experiment_id = exp.run_basic_test_with_temp_config(
+        config, conf.fixtures_path("estimator_dataset"), 1
+    )
+
+    trials = exp.experiment_trials(experiment_id)
```
Comment (aaron276h) [inline e2e_tests/tests/experiment/test_tf_estimator.py:166]: non-blocking: just make this: `trial = exp.experiment_trials(experiment_id)[0]`

File: harness/determined/estimator/_estimator_context.py (hunk 6 above)
Comment (aaron276h) [reply harness/determined/estimator/_estimator_context.py]: Yeah i like the current version, left one comment relate to the last paragraph.

File: e2e_tests/tests/experiment/test_tf_estimator.py (hunk 12 above)
Comment (rb-determined-ai) [reply e2e_tests/tests/experiment/test_tf_estimator.py:159]: good catch.  I was testing on 2 GPUs, and I thought I updated it back to 8, but I guess not.

File: e2e_tests/tests/fixtures/estimator_dataset/model.py (hunk 8 above)
Comment (rb-determined-ai) [reply e2e_tests/tests/fixtures/estimator_dataset/model.py:61]: sure, done

File: harness/determined/estimator/_estimator_context.py (hunk 9 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_estimator_context.py:205]: Cool, rewrote this for better clarity.

File: harness/determined/estimator/_reducer.py (hunk 10 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_reducer.py:10]: oops, definitely yes.  Added.

File: harness/tests/experiment/fixtures/estimator_linear_model.py (hunk 11 above)
Comment (rb-determined-ai) [reply harness/tests/experiment/fixtures/estimator_linear_model.py:43]: thank you!

File: e2e_tests/tests/experiment/test_tf_estimator.py (hunk 13 above)
Comment (rb-determined-ai) [reply e2e_tests/tests/experiment/test_tf_estimator.py:166]: Done.

File: harness/determined/estimator/_reducer.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        self._det_last_accumulate = None
+
+        if isinstance(numpy_dtype, tf.dtypes.DType):
+            raise TypeError(f"numpy_dtype parameter must not be a TensorFlow dtype: {numpy_dtype}")
+        self._det_np_dtype = np.dtype(numpy_dtype)
+        self._det_tf_dtype = tf.compat.v1.as_dtype(numpy_dtype)
+
+        self.update_state(metric)
+
+        @self._det_context._build_allgather_op
```
Comment (aaron276h) [inline harness/determined/estimator/_reducer.py:146]: question: why make this a decorator and not just a function call?

File: docs/reference/api/estimator.txt [hunk 15]
```diff
@@ -44,7 +44,23 @@ API or Native API.
...
 .. autoclass:: determined.estimator.EstimatorExperimentalContext
-    :members: cache_train_dataset, cache_validation_dataset
+    :members: cache_train_dataset, cache_validation_dataset, make_metric
+    :member-order: bysource
+
+
+Reducing Metrics
+~~~~~~~~~~~~~~~~
+
+Determined supports proper reduction of arbitrary metrics during distributed
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply docs/reference/api/estimator.txt]: I made the following edits to the docstrings:
```diff
 Reducing Metrics
 ~~~~~~~~~~~~~~~~

-Determined supports proper reduction of arbitrary metrics during distributed
-training by allowing users to define custom reducers for their metrics. Custom
-reducers can be either a function or an implementation of the
+Determined supports proper reduction of arbitrary validation metrics during
+distributed training by allowing users to define custom reducers for their
+metrics. Custom reducers can be either a function or an implementation of the
```
```diff
     def make_metric(..) -> tf.keras.metrics.Metric:
         """
-        Return an estimator-compatible metric which will be calculated properly, even during
-        distributed training.
+        Return an estimator-compatible validation metric which will be calculated properly, even
+        during distributed evaluation.
```
```diff
 class MetricReducer:
     """
-    Efficiently aggregating metrics across a multi-slot distributed evaluation is done in two steps:
+    Efficiently aggregating validation metrics across a multi-slot distributed evaluation is done
+    in two steps:
```

File: harness/determined/estimator/_reducer.py (hunk 14 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_reducer.py]: Talked offline, the question was about calling `_build_allgather_op` directly rather than applying it as a decorator.  I don't care either way, so I took the direct call approach.
